### PR TITLE
[Interactive graph editor] Add showPointLabels toggle

### PR DESCRIPTION
**Pull request by EmiliaPalaghita**
## Summary

Adds a flag-gated "Show point labels" toggle to the interactive-graph editor that writes `showPointLabels` onto the graph JSON. Uses the `perseus-enable-point-label-field` feature flag.

The toggle is implemented using a `LabeledSwitch` and `InfoTip`. The toggle is disabled until every point has a non-empty label. The new UI is hooked into the `interactive-graph-editor`, above the "Start coordinates" section. 

When the interactive graph is static, the toggle disappears. 

New strings are added in `strings.ts`.

## Demo

https://github.com/user-attachments/assets/7c544777-fffe-46ec-bf4b-a7f18d2c12c6

## Test plan

- pnpm tsc
- pnpm lint
- pnpm test
- Manual Storybook verification of all graph types with the flag on

Issue: [AITQ-385](https://khanacademy.atlassian.net/browse/AITQ-385)

[AITQ-385]: https://khanacademy.atlassian.net/browse/AITQ-385?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx [hunk 1]
```diff
@@ -207,6 +208,29 @@ class InteractiveGraphEditor extends React.Component<Props> {
         }
     };
 
+    changeShowPointLabels = (showPointLabels: boolean) => {
+        const {graph, correct} = this.props;
+        if (!graph?.type || graph.type === "vector" || graph.type === "none") {
+            return;
+        }
+        if (
+            !correct?.type ||
```
Comment (ivyolamit) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx]: Can you please add a comment here that custom point label is not applicable for vector and none graph types

File: packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx [hunk 2]
```diff
@@ -207,6 +208,29 @@ class InteractiveGraphEditor extends React.Component<Props> {
...
+            correct.type === "vector" ||
+            correct.type === "none"
+        ) {
+            return;
+        }
+        // Cast: the runtime guards above narrow to variants that accept
+        // `showPointLabels`, but TS can't preserve that narrowing across a
+        // spread into a discriminated union.
```
Comment (ivyolamit) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx]: No need for this comment here

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx [hunk 3]
```diff
@@ -248,6 +284,30 @@ const StartCoordsSettings = (props: Props) => {
                     {/* Start coordinates input */}
                     <StartCoordsSettingsInner {...props} />
 
+                    {showToggle && (
+                        <View className={styles.switchRow}>
```
Comment (ivyolamit) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: We actually have a component for this `LabeledSwitch`
see `packages/perseus-editor/src/widgets/radio/editor.tsx` or `packages/perseus-editor/src/widgets/image-editor/components/decorative-toggle.tsx` for sample.

Also let's move the InfoTip beside the label similar to this, I'm currently doing a work to make the editor look consistent.
Image

**Review by ivyolamit — APPROVED** [review]
Code looks logical to me, I just have a non-blocking comments. Also I was not able to properly smoke test the feature since it's not fully hooked up to the showPointLabels logic but I can see the toggle showing up.

Image

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx [hunk 4]
```diff
@@ -39,12 +44,16 @@ import type {StartCoords} from "./types";
...
 
+const {InfoTip} = components;
+
 interface StartCoordsSettingsProps {
     range: [x: Range, y: Range];
     step: [x: number, y: number];
     allowReflexAngles?: boolean;
     onChange: (startCoords: StartCoords) => void;
     onChangePointLabels: (pointLabels: ReadonlyArray<string>) => void;
+    showPointLabelsFeatureEnabled?: boolean;
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: Should this line have a TODO comment for removing the feature flag?

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx [hunk 5]
```diff
@@ -228,9 +237,36 @@ const StartCoordsSettingsInner = (props: Props) => {
...
+    const {
+        range,
+        step,
+        onChange,
+        type,
+        showPointLabelsFeatureEnabled,
+        onChangeShowPointLabels,
+    } = props;
     const [isOpen, setIsOpen] = React.useState(true);
+    const switchId = React.useId();
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: The variable name should probably mention which UI element this is for.

```suggestion
    const showPointLabelsSwitchId = React.useId();
```

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx [hunk 6]
```diff
@@ -228,9 +237,36 @@ const StartCoordsSettingsInner = (props: Props) => {
...
+    const {strings} = usePerseusI18n();
+
+    // The toggle is hidden when the feature flag is off, or when the graph
+    // type doesn't support point labels (vector). Other graph types reaching
+    // this component all support `pointLabels` / `showPointLabels`.
+    const showToggle =
+        showPointLabelsFeatureEnabled === true &&
+        type !== "vector" &&
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: `none`-type graphs also don't have point labels, right?

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx [hunk 7]
```diff
@@ -228,9 +237,36 @@ const StartCoordsSettingsInner = (props: Props) => {
     }
 };
 
+const hasPopulatedPointLabels = (props: Props): boolean => {
+    if (!("pointLabels" in props) || !props.pointLabels) {
+        return false;
+    }
+    return props.pointLabels.every((label) => label.trim() !== "");
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: I think we probably want


```suggestion
    return props.pointLabels.some((label) => label.trim() !== "");
```

Otherwise, it will not be possible to show visible labels for graphs where only some of the points are labeled.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx:1]: I don't understand why the setting to show visible labels is being added to the StartCoordsSettings component. What do labels have to do with starting coordinates? If a graph does not have customizable starting coordinates, it should still be possible to add visible labels.

In particular, note that the [StartCoordsSettings are not shown if the graph is `static`](https://github.com/Khan/perseus/blob/a3396604a727b43eabcfb2771704d808b38c14c3/packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/util.ts#L283-L285) (i.e. non-interactive). We want to be able to add visible labels to static graphs.

It seems like we should have another component for controlling whether visible labels are shown, which is used directly in `InteractiveGraphEditor`.

**Review by benchristel — DISMISSED** [review]
I don't think the visible label switch should be added to `StartCoordsSettings`. That makes the switch unavailable when the graph is `static`.

I think we also want to show the toggle when _any_ of the points has a label.

I left some other comments inline.

**Comment by EmiliaPalaghita** [comment]
@benchristel we don't want to show any labels if only part of them are provided, that's how the feature has been implemented. The Perseus JSON for labels is just a list, we don't have a label associated with each coordinate pair, so these changes would require massive changes to Perseus. 

@ivyolamit is this something that would be wanted?

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx (hunk 3 above)
Comment (EmiliaPalaghita) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: Oh that's amazing! Used that instead. Thank you!

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx (hunk 6 above)
Comment (EmiliaPalaghita) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: Yes, it is already gated `packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx` with the new changes.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx (hunk 7 above)
Comment (EmiliaPalaghita) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx]: Due to us wanting to show the labels only when all points have them set, this change is not needed.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx
Comment (EmiliaPalaghita) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-settings.tsx:1]: Oh, that was an oversight on my part! I think I didn't realize the best place to place them and thought it makes sense for the toggle to be next to coordinates. Now it's been moved in the interactive graph settings. Thank you for finding the bug for `static`! That's definitely not what we want.

**Comment by EmiliaPalaghita** [comment]
Updated the demo as well in the description to reflect the new changes.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    showPointLabels: boolean;
+    pointLabels: ReadonlyArray<string> | undefined;
+    onChange: (showPointLabels: boolean) => void;
+}
+
+// Editor-level toggle for the graph's `showPointLabels` field. Rendered at
+// the top of the interactive-graph editor so it stays visible for static
+// graphs, where the start-coords sub-panel (which owns the per-point name
+// inputs) is hidden. The toggle is disabled until every point has a
+// non-empty label; the InfoTip explains why.
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx]: This code has no relation to the start-coords component anymore, so the comment doesn't need to mention it:


```suggestion
// Editor-level toggle for the graph's `showPointLabels` field. The toggle
// is disabled until every point has a non-empty label; the InfoTip
// explains why.
```

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+export default function ShowPointLabelsToggle({
+    showPointLabels,
+    pointLabels,
+    onChange,
+}: Props) {
+    const {strings} = usePerseusI18n();
+    const labelsPopulated =
+        pointLabels !== undefined &&
+        pointLabels.length > 0 &&
+        pointLabels.every((label) => label.trim() !== "");
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx]: This check uses different criteria for label emptiness than other code that handles pointLabels. Elsewhere (e.g. when setting aria-labels on the points) we only consider a label to be absent if it is exactly `undefined` or `""`; we don't consider whitespace-only labels to be empty. I'm not sure if this matters, but thought I'd point it out.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.test.tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        );
+
+        // Assert
+        expect(screen.getByLabelText("Show point labels")).toHaveAttribute(
+            "aria-disabled",
+            "true",
+        );
+    });
+
+    it("disables the toggle when every pointLabels entry is empty", () => {
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.test.tsx:40]: I think we should also have a test for the case where some pointLabels entries are empty and some are not, since we've had conversations about how to handle that specific case.

```ts
it("disables the toggle when some pointLabels are empty and some are not")
```

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx (hunk 8 above)
Comment (benchristel) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx]: Also, the InfoTip doesn't really explain _why_ this logic is used, only _that_ it is used. I think either this comment or the InfoTip should be updated to contain the explanation.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    return (
+        <View className={styles.switchRow}>
+            <LabeledSwitch
+                label={strings.interactiveGraphShowPointLabels}
+                checked={labelsPopulated && showPointLabels}
+                disabled={!labelsPopulated}
+                onChange={onChange}
+            />
+            <InfoTip>{strings.interactiveGraphShowPointLabelsInfoTip}</InfoTip>
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx]: I don't think we usually translate strings that only appear in the editor UI. Compare, for instance, https://github.com/Khan/perseus/blob/af7027bb2f5bb06a73880cba031759bedb6d0196/packages/perseus-editor/src/widgets/interactive-graph-editor/components/interactive-graph-description.tsx#L37-L46. We could hardcode the string here and save translators some time.


```suggestion
            <InfoTip>
                When on, each movable point displays a visible label next to it.
                Add a name to every point below to enable this option.
            </InfoTip>
```

File: packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-shared.module.css [hunk 12]
```diff
@@ -72,3 +72,11 @@
...
     margin-top: var(--wb-sizing-size_120) !important;
 }
+
+.switch-row {
+    display: flex !important;
+    flex-direction: row !important;
+    align-items: center !important;
+    gap: var(--wb-sizing-size_080) !important;
+    margin-top: var(--wb-sizing-size_120) !important;
+}
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/start-coords/start-coords-shared.module.css]: Now that `show-point-labels-toggle.tsx` is its own component, the styles for it should live in a corresponding CSS file — not here in the `start-coords` CSS.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx [hunk 13]
```diff
@@ -207,6 +209,28 @@ class InteractiveGraphEditor extends React.Component<Props> {
...
+            correct.type === "vector" ||
+            correct.type === "none"
+        ) {
+            return;
+        }
+        this.props.onChange({
+            // eslint-disable-next-line no-restricted-syntax
+            graph: {...graph, showPointLabels} as PerseusGraphType,
+            // eslint-disable-next-line no-restricted-syntax
+            correct: {...correct, showPointLabels} as PerseusGraphType,
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx]: If, in the future, we add another graph type that doesn't have point labels, these `as` casts will suppress the type error and we'll probably forget to update the `if` statement above that checks the graph type.

The typesafe way to write code that differs by graph type is with a switch statement:

```ts
if (!graph || !correct) {
  return;
}

let newGraph: PerseusGraphType;
let newCorrect: PerseusGraphType;
switch (graph.type) {
  case "absolute-value":
    newGraph = {...graph, showPointLabels};
    newCorrect = {...correct, showPointLabels};
    break;
  case "angle":
    newGraph = {...graph, showPointLabels};
    newCorrect = {...correct, showPointLabels};
    break;
  // ...
  case "none":
    return;
  // ...
  default:
    // Make TS prove the switch is exhaustive.
    throw new UnreachableCaseError(graph);
}
```

This verbosity is unfortunately necessary for TypeScript to figure out what's going on :(

We're trying to move in the direction of more type safety in the Perseus editors, since it is currently a blind spot for us where bugs can easily creep in.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx [hunk 14]
```diff
@@ -512,6 +543,31 @@ class InteractiveGraphEditor extends React.Component<Props> {
...
+                            !this.props.static &&
+                            this.props.graph?.type &&
+                            this.props.graph.type !== "vector" &&
+                            this.props.graph.type !== "none" && (
+                                <ShowPointLabelsToggle
+                                    showPointLabels={
+                                        "showPointLabels" in this.props.graph
+                                            ? this.props.graph
+                                                  .showPointLabels === true
+                                            : false
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx]: Since we checked the graph type above, we don't also need to check for `"showPointLabels" in this.props.graph` here.


```suggestion
                                        this.props.graph.showPointLabels === true
```

File: packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx [hunk 15]
```diff
@@ -512,6 +543,31 @@ class InteractiveGraphEditor extends React.Component<Props> {
...
+                                    }
+                                    pointLabels={
+                                        "pointLabels" in this.props.graph
+                                            ? this.props.graph.pointLabels
+                                            : undefined
```
Comment (benchristel) [inline packages/perseus-editor/src/widgets/interactive-graph-editor/interactive-graph-editor.tsx]: Same here.


```suggestion
                                        this.props.graph.pointLabels
```

File: packages/perseus/src/widgets/interactive-graphs/mafs-graph.test.tsx [hunk 16]
```diff
@@ -1387,6 +1387,29 @@ describe("MafsGraph", () => {
                 screen.queryByTestId("movable-point__visible-label"),
             ).not.toBeInTheDocument();
         });
+
+        it("does not show the toggle for static graphs, when flag on and showPointLabels: true", () => {
```
Comment (benchristel) [inline packages/perseus/src/widgets/interactive-graphs/mafs-graph.test.tsx:1391]: ~~I don't think this test is correct. Static graphs have the "movable" point elements, but they are grayed out and not interactive. There's no technical reason we can't add labels to static graphs. Is there a product reason?~~

For reference, here's an example of a static graph: https://khan.github.io/perseus/?path=/story/widgets-interactive-graph-visual-regression-tests-initial-state--static-graph

~~The original use of static graphs was for hints in exercises, where the hint needs to show how a learner might position the interactive elements of the graph. I can see labels being useful — or even required — for this use case.~~

EDIT: I think my suggested change can be a future issue/PR. No changes needed here.

File: packages/perseus/src/widgets/interactive-graphs/mafs-graph.tsx [hunk 17]
```diff
@@ -396,7 +396,10 @@ export const MafsGraph = (props: MafsGraphProps) => {
                         {isFeatureOn(
                             {apiOptions: props.apiOptions},
                             "perseus-enable-point-label-field", // TODO(AITQ-385): clean up feature flag
-                        ) && <MovablePointLabelsLayer state={state} />}
+                        ) &&
+                            !props.static && (
+                                <MovablePointLabelsLayer state={state} />
+                            )}
```
Comment (benchristel) [inline packages/perseus/src/widgets/interactive-graphs/mafs-graph.tsx:402]: ~~Same reasoning as the test above — I think this can be~~


```suggestion
                        ) && <MovablePointLabelsLayer state={state} />
```

~~One of the reasons I wanted the "show labels" toggle moved out of the StartCoordsSettings component (besides code organization) was so that it would be shown when the graph was static.~~

EDIT: sorry, please disregard this comment — I see now why it makes sense to not show labels for static graphs (they are not editable).

File: packages/perseus/src/strings.ts [hunk 18]
```diff
@@ -634,6 +634,8 @@ export type PerseusStrings = {
     gifPlayButtonLabel: string;
     gifPauseButtonLabel: string;
     definitionIdentifier: ({word}: {word: string}) => string;
+    interactiveGraphShowPointLabels: string;
+    interactiveGraphShowPointLabelsInfoTip: string;
```
Comment (benchristel) [inline packages/perseus/src/strings.ts]: Re: my comment above — I don't think we need translatable strings for these labels.

**Review by benchristel — COMMENTED** [review]
~~We're now not showing visible point labels for static graphs — is that desired? In my last review, I mentioned that we should _show_ the point labels switch in the editor when the graph is static. It seems to me that static graphs should be allowed to have point labels, but if that's wrong, let me know!~~

~~From my inline comment:~~

> ~~The original use of static graphs was for hints in exercises, where the hint needs to show how a learner might position the interactive elements of the graph. I can see labels being useful — or even required — for this use case.~~

EDIT: Ivy pointed out that the point labels themselves are edited via the start coords settings, so they're not available for static graphs. In light of that, I agree that we should disable the "show point labels" switch when the graph is static — at least for now.

I requested some other changes inline.

**Comment by ivyolamit** [comment]
> We're now not showing visible point labels for static graphs — is that desired? In my last review, I mentioned that we should _show_ the point labels switch in the editor when the graph is static. It seems to me that static graphs should be allowed to have point labels, but if that's wrong, let me know!
> 
> From my inline comment:
> 
> > The original use of static graphs was for hints in exercises, where the hint needs to show how a learner might position the interactive elements of the graph. I can see labels being useful — or even required — for this use case.
> 
> I requested some other changes inline.

@benchristel the static part is something I miss 😅 
The reason I suggested it below the Start Coordinates it's because the label input is also there and make it more visible for content editor when they add the custom point label. But I don't have strong opinion where it should be placed.

File: packages/perseus/src/strings.ts (hunk 18 above)
Comment (ivyolamit) [reply packages/perseus/src/strings.ts]: This is something that was discussed in the team before, and I was told that we also add translation for our editor. ~~but i might be wrong...let me find the discussion thread...~~ ok i'm wrong about translating the actual text in the editor 🙈 sorry.

**Comment by benchristel** [comment]
@ivyolamit thanks for that additional context! I just realized that when the graph is static, content creators have no way to change the labels, since the start coords UI is not available.

I think this is a problem for accessibility of existing static graphs 😕.

I wonder if the point label settings should be moved out of the start coords UI, so the labels can be set for static graphs. That would increase the scope of this change, though.

For now, I think I am fine with the following:

- disable the `show point labels` switch and hide the labels when the graph is static (this PR already does that — no change needed!)
- create a Jira ticket to make point labels editable for static graphs.

@EmiliaPalaghita

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx (hunk 9 above)
Comment (EmiliaPalaghita) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx]: Thanks for pointing it out! For now I removed `trim`. I think we would ideally want to handle this edge case too, including others that might not be obvious.

File: packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx (hunk 11 above)
Comment (EmiliaPalaghita) [reply packages/perseus-editor/src/widgets/interactive-graph-editor/components/show-point-labels-toggle.tsx]: @ivyolamit suggested to add strings to `strings.ts`, so that's why I added them there. I didn't know the Editor facing text is hardcoded. I updated accordingly.
